File: primalTestDotNet/AI/AIStateMachine.cs
using PrimalTestDotNet.AI.States;
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI;

public class AIStateMachine
{
    private IState _state;
    private readonly Level.Level _level;
    private readonly Hero _hero;
    private readonly List<IGameObject> _gameObjects;

    public IState LeaveLevelState => new LeavingLevelState(this, _level);
    public IState UnwinnableState => new UnwinnableState(_level);
    public IState LookingForPotionState => new LookingForPotionState(this, _level);
    public IState LookingForSwordState => new LookingForSwordState(this, _level);
    public IState LookingForTreasureState => new LookingForTreasureState(this, _level);

    public AIStateMachine(Level.Level level, Hero hero, List<IGameObject> gameObjects)
    {
        _state = new LookingForTreasureState(this, level);

        _level = level;
        _hero = hero;
        _gameObjects = gameObjects;
    }

    public void ChangeState(IState state)
    {
        _state = state;
    }

    public void Update()
    {
        _state.Update(_hero, _gameObjects);
    }
}

File: PrimalTestCpp/AIStateMachine.cpp
#include "AIStateMachine.h"
#include "IState.h"

AIStateMachine::AIStateMachine(Level& level, Hero& hero, std::list<std::shared_ptr<IGameObject>>& gameObjects)
	: _level(level), _hero(hero), _gameObjects(gameObjects)
{
	_state = LOOKING_FOR_TREASURE_STATE();
}

void AIStateMachine::changeState(std::shared_ptr<IState> state) {
	_state = std::move(state);
}

void AIStateMachine::update() {
	_state->update(_hero, _gameObjects);
}
